ansible_base/lib/utils/db.py

```diff
@@ -25,3 +26,126 @@ def migrations_are_complete() -> bool:
     executor = MigrationExecutor(connection)
     plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
     return not bool(plan)
+
+
+# NOTE: the django_pglocks_advisory_lock context manager was forked from the django-pglocks v1.0.4
+# that was licensed under the MIT license
+
+
+@contextmanager
+def django_pglocks_advisory_lock(lock_id, shared=False, wait=True, using=None):
+
+    if using is None:
+        using = DEFAULT_DB_ALIAS
```

Comment:
This comes from the library:

https://github.com/Xof/django-pglocks/blob/master/django_pglocks/__init__.py#L10

And the reason is clearly because of a hesitancy to import from `django.db` at the top level of the file. For many imports this is a valid concern, due to the "apps are not loaded yet" error. But DAB imports from paths in Django in many places in a way that relies on those specific paths not requiring app setup in advance. So I agree, we could move the import to the top of the file and make that the default.

Reply:
If you are feeling frisky you can use `get_connection()` and dodge reaching for a possibly not stable constant. But that doesn't solve the underlying import hairyness, it also deviates from the copy paste.
 
https://github.com/django/django/blob/3ee4c6a27ad520d4ecc3cded260d47cbccafb144/django/db/transaction.py#L18

Reply:
It's not quite the same, as `get_connection` returns the wrapper for the connection, but `connection` returns a proxy for the wrapper.

```
In [6]: connection
Out[6]: <django.utils.connection.ConnectionProxy at 0x7fd28ddc4530>

In [7]: get_connection()
Out[7]: <DatabaseWrapper vendor='sqlite' alias='default'>
```

But if you look at the code for `ConnectionProxy`, its life goal is to pretend to be the thing it wraps (the `DatabaseWrapper`), while still allowing configuration changes to change out the specific `DatabaseWrapper` to change out from under it.

So end then end, they are different, but as long as you are only referencing its properties (like `connection.cursor`), it shouldn't matter. It is designed to not matter.

Reply:
Filed

https://github.com/ansible/django-ansible-base/issues/679

I do generally agree, but we'd need to approach this incrementally.